Support inequal join for `all` `left/right/full/inner`

lgbo-ustc opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature



### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Support join with inequal conditions which involve columns from both left and right table. e.g. `t1.y < t2.y`. To enable, `SET allow_experimental_join_condition = 1`

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

lgbo-ustc committed:
support inequal join for left/right/inner all

lgbo-ustc commented:
Have some tests.

Run the filter condition outside join, which would filtout most of the data
```sql
f2386dc7dd0d :) select * from (select t1.y as y1 , t2.y as y2, t1.*, t2.* from test_join_1 as t1 inner join test_join_2 as t2 on t1.x = t2.x) where y1 < y2 Format Null settings join_algorithm='hash', max_threads=1, allow_experimental_analyzer=1;


SELECT *
FROM
(
    SELECT
        t1.y AS y1,
        t2.y AS y2,
        t1.*,
        t2.*
    FROM test_join_1 AS t1
    INNER JOIN test_join_2 AS t2 ON t1.x = t2.x
)
WHERE y1 < y2
FORMAT `Null`
SETTINGS join_algorithm = 'hash', max_threads = 1, allow_experimental_analyzer = 1

Query id: 3a72f93b-8916-48df-80fc-f487a61e3878

Ok.

0 rows in set. Elapsed: 7.968 sec. Processed 400.00 thousand rows, 51.60 MB (50.20 thousand rows/s., 6.48 MB/s.)
```
run on the same data set, but move the filter into join
```
f2386dc7dd0d :) select * from (select t1.y as y1 , t2.y as y2, t1.*, t2.* from test_join_1 as t1 inner join test_join_2 as t2 on t1.x = t2.x) where y1 < y2 Format Null settings join_algorithm='hash', max_threads=1, allow_experimental_analyzer=1;


SELECT *
FROM
(
    SELECT
        t1.y AS y1,
        t2.y AS y2,
        t1.*,
        t2.*
    FROM test_join_1 AS t1
    INNER JOIN test_join_2 AS t2 ON t1.x = t2.x
)
WHERE y1 < y2
FORMAT `Null`
SETTINGS join_algorithm = 'hash', max_threads = 1, allow_experimental_analyzer = 1

Query id: 3a72f93b-8916-48df-80fc-f487a61e3878

Ok.

0 rows in set. Elapsed: 7.968 sec. Processed 400.00 thousand rows, 51.60 MB (50.20 thousand rows/s., 6.48 MB/s.)
```

run on a data set which will not filtout any row
```
f2386dc7dd0d :) select t1.*, t2.* from test_join_1 as t1 inner join test_join_2 as t2 on t1.x = t2.x and t1.y < t2.y Format Null settings join_algorithm='hash', max_threads=1, allow_experimental_analyzer=1;

SELECT
    t1.*,
    t2.*
FROM test_join_1 AS t1
INNER JOIN test_join_2 AS t2 ON (t1.x = t2.x) AND (t1.y < t2.y)
FORMAT `Null`
SETTINGS join_algorithm = 'hash', max_threads = 1, allow_experimental_analyzer = 1

Query id: 4c5cdb4c-e9f7-4d86-8945-693076511aea

Ok.

0 rows in set. Elapsed: 4.687 sec. Processed 400.00 thousand rows, 51.60 MB (85.34 thousand rows/s., 11.01 MB/s.)
Peak memory usage: 55.03 MiB.
```

vdimir commented:
Looks promising, thanks for working on that!
My previous attempt https://github.com/ClickHouse/ClickHouse/pull/46479 , you may want to take the test from it

lgbo-ustc commented:
> Looks promising, thanks for working on that! My previous attempt #46479 , you may want to take the test from it

OK.
Part of the code in this PR is based on #46479

lgbo-ustc committed:
fixed bugs

lgbo-ustc committed:
Optimization for cutting left blocks into processed and not_processed parts.

lgbo-ustc committed:
fixed bugs for all right join

add test cases

fixed typos

add document

fixed stype error and rename some variables

fixed test error

add test cases

try to recude hash join code size

debug

fixed

fixed coredump

fixed. condition result type is not bool

enable when allow_experimental_analyzer=0

fixed tests

fixed

update tests

update tests

roll backup when use the old analyzer

fixed typos

fixed

test

large obj check

test

test

lgbo-ustc commented:
@vdimir  The code of `HashJoin` seems too be large. Is there any way to solve?
```
/build/src/Interpreters/HashJoin.cpp
Mar 08 08:42:14 1.	<eof> parser at end of file
Mar 08 08:42:14 2.	Code generation
Mar 08 08:42:14 LLVM ERROR: out of memory
Mar 08 08:42:14 Allocation failed
Mar 08 08:42:14 sccache: Compiler killed by signal 6
```

vdimir commented:
>  The code of HashJoin seems too large. Is there any way to solve?

I believe some template arguments are pointless and can be replaced with flags, so we may try to find ones that are not used in hot loop

lgbo-ustc commented:
https://s3.amazonaws.com/clickhouse-test-reports/60920/e47fd43fd32795ffe3d416f87fb8473fb41454f3/fast_test.html

lgbo-ustc commented:
@vdimir could you help to check if there is any other problem with the CI process ? fast test keep failing on some unrelated tests

vdimir commented:
@lgbo-ustc it can happen in case when server crashed, other tests will marked as failed due to server not responded.

It's indeed not really convenient to find reason in such cases, but you may find failure info in server logs (usually message is `Tests are not finished` indicates that):

For example for https://github.com/ClickHouse/ClickHouse/pull/60920/commits/7e5d43457cf083e9a394f79c4bd8beb185c6f09c server failed with
`2024.03.11 11:31:32.360388 [ 219775 ] {} <Fatal> BaseDaemon: (version 24.3.1.1, build id: 4463762629B619462AE487D79A0FF83DFA2DDC50, git hash: 9b9275601c6bb4e2e9447d9f4460fb6075108a76) (from thread 190031) (query_id: 79f68eba-461e-418c-ace9-e5758e33740d) (query: SELECT t1.*, t2.* FROM t1 LEFT JOIN t2 ON (t1.a < t2.a OR lower(t1.attr) == lower(t2.attr)) AND t1.key = t2.key ORDER BY (t1.key, t1.attr, t2.key, t2.attr);) Received signal Segmentation fault (11)`

<details>
<summary>Details</summary>


```
~ % curl -s https://s3.amazonaws.com/clickhouse-test-reports/60920/7e5d43457cf083e9a394f79c4bd8beb185c6f09c/fast_test/server.log.zst | zstd -d | grep  '<Fatal>' 
2024.03.11 11:31:32.360250 [ 219775 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.03.11 11:31:32.360285 [ 219775 ] {} <Fatal> BaseDaemon: (version 24.3.1.1, build id: 4463762629B619462AE487D79A0FF83DFA2DDC50, git hash: 9b9275601c6bb4e2e9447d9f4460fb6075108a76) (from thread 190031) Received signal 11
2024.03.11 11:31:32.360310 [ 219775 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
2024.03.11 11:31:32.360321 [ 219775 ] {} <Fatal> BaseDaemon: Address: 0xb1. Access: write. Address not mapped to object.
2024.03.11 11:31:32.360351 [ 219775 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000078a3588 0x0000000007b0fb7f
2024.03.11 11:31:32.360361 [ 219775 ] {} <Fatal> BaseDaemon: ########################################
2024.03.11 11:31:32.360388 [ 219775 ] {} <Fatal> BaseDaemon: (version 24.3.1.1, build id: 4463762629B619462AE487D79A0FF83DFA2DDC50, git hash: 9b9275601c6bb4e2e9447d9f4460fb6075108a76) (from thread 190031) (query_id: 79f68eba-461e-418c-ace9-e5758e33740d) (query: SELECT t1.*, t2.* FROM t1 LEFT JOIN t2 ON (t1.a < t2.a OR lower(t1.attr) == lower(t2.attr)) AND t1.key = t2.key ORDER BY (t1.key, t1.attr, t2.key, t2.attr);) Received signal Segmentation fault (11)
2024.03.11 11:31:32.360405 [ 219775 ] {} <Fatal> BaseDaemon: Address: 0xb1. Access: write. Address not mapped to object.
2024.03.11 11:31:32.360419 [ 219775 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000078a3588 0x0000000007b0fb7f
2024.03.11 11:31:32.372254 [ 219775 ] {} <Fatal> BaseDaemon: 0.0. inlined from /fasttest-workspace/build/./src/Common/StackTrace.cpp:349: StackTrace::tryCapture()
2024.03.11 11:31:32.372290 [ 219775 ] {} <Fatal> BaseDaemon: 0. /fasttest-workspace/build/./src/Common/StackTrace.cpp:318: StackTrace::StackTrace(ucontext const&) @ 0x00000000078a3588
2024.03.11 11:31:32.399012 [ 219775 ] {} <Fatal> BaseDaemon: 1. /fasttest-workspace/build/./src/Daemon/BaseDaemon.cpp:0: signalHandler(int, siginfo_t*, void*) @ 0x0000000007b0fb7f
2024.03.11 11:31:32.406064 [ 219775 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.03.11 11:31:32.441672 [ 219775 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
2024.03.11 11:31:32.441751 [ 219775 ] {} <Fatal> BaseDaemon: Changed settings: connect_timeout_with_failover_ms = 2000, connect_timeout_with_failover_secure_ms = 3000, idle_connection_timeout = 36000, s3_check_objects_after_upload = true, stream_like_engine_allow_direct_select = true, replication_wait_for_inactive_replica_timeout = 30, allow_nonconst_timezone_arguments = true, log_queries = true, insert_quorum_timeout = 60000, fsync_metadata = false, http_send_timeout = 60., http_receive_timeout = 60., opentelemetry_start_trace_probability = 0.10000000149011612, allow_experimental_analyzer = true, join_algorithm = 'hash', max_untracked_memory = 1048576, memory_profiler_step = 1048576, log_comment = '03006_join_on_inequal_expression_short.gen.sql', send_logs_level = 'fatal', allow_introspection_functions = true, database_atomic_wait_for_drop_and_detach_synchronously = true, distributed_ddl_entry_format_version = 6, async_insert_busy_timeout_max_ms = 5000, load_marks_asynchronously = true, allow_prefetched_read_pool_for_remote_filesystem = false, allow_prefetched_read_pool_for_local_filesystem = false, filesystem_prefetch_max_memory_usage = 1073741824, insert_keeper_max_retries = 100, insert_keeper_retry_initial_backoff_ms = 1, insert_keeper_retry_max_backoff_ms = 1, insert_keeper_fault_injection_probability = 0.009999999776482582
2024.03.11 11:31:36.860098 [ 158499 ] {} <Fatal> Application: Child process was terminated by signal 11.
```


</details>

lgbo-ustc commented:
> @lgbo-ustc it can happen in case when server crashed, other tests will marked as failed due to server not responded.
> 
> It's indeed not really convenient to find reason in such cases, but you may find failure info in server logs (usually message is `Tests are not finished` indicates that):
> 
> For example for [7e5d434](https://github.com/ClickHouse/ClickHouse/commit/7e5d43457cf083e9a394f79c4bd8beb185c6f09c) server failed with `2024.03.11 11:31:32.360388 [ 219775 ] {} <Fatal> BaseDaemon: (version 24.3.1.1, build id: 4463762629B619462AE487D79A0FF83DFA2DDC50, git hash: 9b9275601c6bb4e2e9447d9f4460fb6075108a76) (from thread 190031) (query_id: 79f68eba-461e-418c-ace9-e5758e33740d) (query: SELECT t1.*, t2.* FROM t1 LEFT JOIN t2 ON (t1.a < t2.a OR lower(t1.attr) == lower(t2.attr)) AND t1.key = t2.key ORDER BY (t1.key, t1.attr, t2.key, t2.attr);) Received signal Segmentation fault (11)`
> 
> Details

Thanks a lot

lgbo-ustc commented:
@vdimir There are existing tests which run with `allow_experimental_analyzer=1` automatically. To make these tests pass, the older analyzer also need to be modified. But I found that this will be a big change, cannot ensure it could behavior fine.

Or should I add a setting to control whether to enable the new features. What's your opinion.

vdimir commented:
@lgbo-ustc it's alright to introduce a feature only for a new analyzer, so you can just add `allow_experimental_analyzer=1` to the new tests.

What is the purpose of `enable_mixed_join_condition`? If it doesn't affect existing queries, shouldn't the feature just to be enabled by default?

lgbo-ustc commented:
> @lgbo-ustc it's alright to introduce a feature only for a new analyzer, so you can just add `allow_experimental_analyzer=1` to the new tests.
> 
> What is the purpose of `enable_mixed_join_condition`? If it doesn't affect existing queries, shouldn't the feature just to be enabled by default?

I found the exsisting tests, e.g. `01881_join_on_conditions_hash` will also be executed with `allow_experimental_analyzer` enable, although this flag is not setted in the sql file.

I add `enable_mixed_join_condition ` to avoid these tests be affected by this new feature.

lgbo-ustc commented:
@vdimir the build failed. Are the file too large? 
```
Mar 18 21:23:29 FAILED: src/Functions/CMakeFiles/clickhouse_functions_obj.dir/FunctionsConversion.cpp.o 
Mar 18 21:23:29 sccache: warning: The server looks like it shut down unexpectedly, compiling locally instead
Mar 18 21:23:29 sccache: Compile terminated by signal 9
Mar 18 21:23:29 [9473/9855] Building CXX object src/CMakeFiles/dbms.dir/Interpreters/HashJoin.cpp.o,
```

lgbo-ustc committed:
reduce hash join code size

add check

debug

fixed

fixed

fixed

fixed

fixed

debug

debug

update

rename full_expression

rollback ExecutionContext

experiment: support join_use_nulls

vdimir commented:
> Are the files too large?

Probably yes, but not really clear why it's increased, since not too many templates are added. Do we need introduction `AddedColumns` template argument? Since templates in `AddedColumns` itself are not changed?

lgbo-ustc commented:
> > Are the files too large?
> 
> Probably yes, but not really clear why it's increased, since not too many templates are added. Do we need introduction `AddedColumns` template argument? Since templates in `AddedColumns` itself are not changed?

I move some template flags into function arguments instead without too much performance loss. Let's see if we can pass this time. But can't say about `FunctionsConvert.cpp`

lgbo-ustc commented:
still fail on `FunctionsConversion.cpp`

add an issue #61593

lgbo-ustc commented:
@vdimir  I thinks this is ready, Do you have time to review it?

vdimir commented on src/Core/Block.cpp:
```diff
@@ -430,6 +433,16 @@ std::string Block::dumpIndex() const
     return out.str();
 }
 
+std::string Block::dumpContent() const
```
We have `formatBlock` function. Calling it requires passing `output_format` but it can be easyly retieved, so the same would be 3 lines of code, a bit more, but I don't think it's worth intorudcing such a `dumpContent` methond.

https://github.com/bigo-sg/ClickHouse/blob/216dcbef2c86a58a0c8b3679395e89d3fec640b2/src/Formats/formatBlock.cpp#L11

vdimir commented on src/Interpreters/HashJoin.cpp:
```diff
@@ -1480,9 +1494,329 @@ void setUsed(IColumn::Filter & filter [[maybe_unused]], size_t pos [[maybe_unuse
+            sample_right_block.dumpNames(),
+            added_columns.left_block.dumpNames());
+    }
+    LOG_TRACE(getLogger("HashJoin"), "Additional filter execute block:\n{}", executed_block.dumpContent());
+    added_columns.additional_filter_expression->execute(executed_block);
+    LOG_TRACE(getLogger("HashJoin"), "Addition filter execute result block:\n{}", executed_block.dumpContent());
```
I assume printing blocks is too verbose, so we can remove these logs

vdimir commented on src/Interpreters/HashJoin.cpp:
```diff
@@ -1480,9 +1494,329 @@ void setUsed(IColumn::Filter & filter [[maybe_unused]], size_t pos [[maybe_unuse
+    typename Map,
+    bool need_replication,
+    typename AddedColumns>
+NO_INLINE size_t joinRightColumnsWithAddtitionalFilter(
```
Can't we join blocks using regular `joinRightColumns` and filter afterwards? It's quite complex code and maintaining duplicate version can be tricky

vdimir replied:
What if we add code handling mixedExpression into `joinRightColumns` under `if (mixed_join_expression)` (with boolean template flag and `if constexpr` if it would be some overhead). Is it possible?

vdimir commented on src/Planner/PlannerJoinTree.cpp:
```diff
@@ -1318,6 +1318,14 @@ JoinTreeQueryPlan buildQueryPlanForJoinNode(const QueryTreeNodePtr & join_table_
             }
         }
+
+        if (join_clauses_and_actions.full_join_expressions_actions)
```
Maybe we can rename this expressions and method to `mixed` instead of full to have more unoform naming?

vdimir commented on utils/check-style/check-large-objects.sh:
```diff
@@ -9,7 +9,7 @@ TU_EXCLUDES=(
 )
 
-if find $1 -name '*.o' | xargs wc -c | grep --regexp='\.o$' | sort -rn | awk '{ if ($1 > 50000000) print }' \
+if find $1 -name '*.o' | xargs wc -c | grep --regexp='\.o$' | sort -rn | awk '{ if ($1 > 100000000) print }' \
```
What contribytes to big binary size? Is it `joinRightColumnsWithAddtitionalFilter`. I'm afraind we shouldn't just increase the limit because it can lead to build issues, like we had with `RangeHashedDictionary` recently https://github.com/ClickHouse/ClickHouse/pull/61162

vdimir commented on src/Planner/PlannerJoins.cpp:
```diff
@@ -357,24 +349,58 @@ void buildJoinClause(
         }
+        else
+        {
+            auto support_mixed_join_condition = planner_context->getQueryContext()->getSettingsRef().enable_mixed_join_condition;
```
Is the only purpose of the setting is to throw exception? In that case we can just allow new syntax. Or introduce it as experimental for some time, `allow_experimental_join_condition`

vdimir commented on src/Planner/PlannerJoins.cpp:
```diff
@@ -845,6 +912,13 @@ std::shared_ptr<IJoin> chooseJoinAlgorithm(std::shared_ptr<TableJoin> & table_jo
     const PlannerContextPtr & planner_context)
 {
+    if (table_join->getFullJoinExpression() && !table_join->isEnabledAlgorithm(JoinAlgorithm::HASH)
+        && !(table_join->isEnabledAlgorithm(JoinAlgorithm::GRACE_HASH) && table_join->oneDisjunct()))
```
Grace hash join doesn't support multiple disjuncts anyway
https://github.com/ClickHouse/ClickHouse/blob/15a0f5381f9ec07cb04d339eb836222b628f2a03/src/Planner/PlannerJoins.cpp#L881-L882

vdimir reviewed (commented):
Sorry for delay again.

To be honest, introducing `joinRightColumnsWithAdditionalFilter` is very concerning, since it duplicates code from `joinRightColumns`, which is pretty complicated and blows up a binary. What is the motivation behind this? Is it performance? How much slower is it if we use the existing method to join and then apply a filter? (Also, we need to move non-joined to defaults, maybe that's the trickiest part?)

lgbo-ustc commented:
> Sorry for delay again.
> 
> To be honest, introducing `joinRightColumnsWithAdditionalFilter` is very concerning, since it duplicates code from `joinRightColumns`, which is pretty complicated and blows up a binary. What is the motivation behind this? Is it performance? How much slower is it if we use the existing method to join and then apply a filter? (Also, we need to move non-joined to defaults, maybe that's the trickiest part?)

It's for performance.  In our case
``` sql
select t1.*, t2.* from t1 inner join t2 on t1.key = t2.key and t1.x < t2.x
```
`t1.x < t2.x` filts out most of the data. If we apply the filter after `joinRightColumns`, during `joinRightColumns` there will be a lot of useless data are copied. The runtime will be same as the following sql
```sql
select * from (select t1.*, t2.* from t1 inner join t2 on t1.key = t2.key ) t1.x < t2.x
```
It's an order of magnitude slower

vdimir commented:
> It's for performance. In our case ...

Could you please also give a sample query to benchmark (add to performance test, or just put in a comment here) ?
I'd also like to try to check if it's possible to simplify code, to be sure that we are talking about performance for the same scenario.

lgbo-ustc commented:
> > It's for performance. In our case ...
> 
> Could you please also give a sample query to benchmark (add to performance test, or just put in a comment here) ? I'd also like to try to check if it's possible to simplify code, to be sure that we are talking about performance for the same scenario.

This is a new feature, not sure whether we could a test in the performace test. We test the performance as following
```sql
create table test_join_1(
    `x` UInt64,
    `y` UInt64,
    `s1` String,
    `s2` String,
    `s3` String,
    `s4` String,
    `s5` String,
    `s6` String,
    `s7` String,
    `s8` String,
) ENGINE = Memory;

create table test_join_2(
    `x` UInt64,
    `y` UInt64,
    `k1` String,
    `k2` String,
    `k3` String,
    `k4` String,
    `k5` String,
    `k6` String,
    `k7` String,
    `k8` String,
) ENGINE = Memory;

insert into test_join_1 select number / 200 as x, rand32() % 100 + 90 as y, randomString(6), randomString(5),randomString(5), randomString(5),randomString(5), randomString(5),randomString(5), randomString(5)  from numbers(200000);
insert into test_join_2 select number / 200 as x, rand32() % 100 as y, randomString(6), randomString(5),randomString(5), randomString(5),randomString(5), randomString(5),randomString(5), randomString(5)  from numbers(200000);
select t1.*, t2.* from test_join_1 as t1 inner join test_join_2 as t2 on t1.x = t2.x and t1.y < t2.y Format Null settings join_algorithm='hash', max_threads=1, allow_experimental_analyzer=1, allow_experimental_join_condition=1;
```

zhanglistar commented:
@vdimir It seems that CI says OK now, do you have time to do code review? Thanks.

vdimir commented on src/Interpreters/ExpressionActions.cpp:
```diff
@@ -715,6 +729,7 @@ void ExpressionActions::execute(Block & block, size_t & num_rows, bool dry_run,
     {
         .inputs = block.data,
         .num_rows = num_rows,
+        .expression = *this,
```
Do we really need this backreference? It's used only in a message for a logical error, so it probably isn't worth introducing because someone later could overuse or misuse it.

lgbo-ustc replied:
This is for exception message, could be removed

vdimir commented on src/Planner/PlannerJoins.cpp:
```diff
@@ -357,24 +349,58 @@ void buildJoinClause(
+            {
+                /// expression involves both tables.
+                /// `expr1(left.col1, right.col2) == expr2(left.col3, right.col4)`
+                const auto * node = appendExpression(mixed_dag, join_expression, planner_context, join_node);
```
I found that dag can be incorrect with `join_use_nulls` or when column change thier type, e.g. `t0.x :: Int8` and `t0.x :: UInt8` and expression is `JOIN ON t0.x = t1.x AND t0.x + t1.x > 2`. It happens because here we have original types in columns in `join_expression` as if it supposed to be executed before joining. But we update tyypes of columns for mentioned cases right after columns goes into join, so inside HashJoin `mixed_dag` is executed on top of nullable or casted to common type columns, what leads to logical error.
I tried to fix it, but haven't come up with a proper solution yet. Maybe for such case we need to preserve original columns and pass it into join additionally with changed names and this dag should use that columns.
Or handle it here to build a dag that is aware about type changes.
Or just forbid such cases with `NOT_IMPLEMENTED` for now.

lgbo-ustc replied:
> What if we add code handling mixedExpression into joinRightColumns under if (mixed_join_expression) (with > boolean template flag and if constexpr if it would be some overhead). Is it possible?

Do you mean ?
```c++
template<bool has_mixed_join_expression >
size_t joinRightColumns()
{
    if constexpr(has_mixed_join_expression)
    {
        // new way to handle
    }
    else
    {
       // as before
    }
}
```

lgbo-ustc replied:
😂`join_use_nulls` is used in our case.

lgbo-ustc replied:
it's strange, this case does not fail.
```
0.x :: Int8 and t0.x :: UInt8 and expression is JOIN ON t0.x = t1.x AND t0.x + t1.x > 2
```

lgbo-ustc replied:
Could we add a `IQueryTreePass` to rewrite the query tree before enter `buildJoinClause` ?

zhanglistar replied:
> I found that dag can be incorrect with `join_use_nulls` or when column change thier type, e.g. `t0.x :: Int8` and `t0.x :: UInt8` and expression is `JOIN ON t0.x = t1.x AND t0.x + t1.x > 2`. It happens because here we have original types in columns in `join_expression` as if it supposed to be executed before joining. But we update tyypes of columns for mentioned cases right after columns goes into join, so inside HashJoin `mixed_dag` is executed on top of nullable or casted to common type columns, what leads to logical error. I tried to fix it, but haven't come up with a proper solution yet. Maybe for such case we need to preserve original columns and pass it into join additionally with changed names and this dag should use that columns. Or handle it here to build a dag that is aware about type changes. Or just forbid such cases with `NOT_IMPLEMENTED` for now.

yes, that's a big issue, we have a try.

vdimir replied:
@lgbo-ustc 

> > What if we add code handling mixedExpression into joinRightColumns under if (mixed_join_expression) (with > boolean template flag and if constexpr if it would be some overhead). Is it possible?
> 
> Do you mean ? ...


Only if amount of common code that would be common is significant and `handle mixed condition` is not too big, like. I'm not sure if it's feasible or not, though
 
```c++
template<bool has_mixed_join_expression >
size_t joinRightColumns()
{
/// common code
    if constexpr(has_mixed_join_expression)
    {
        // handle mixed condition
    }
/// common code
}
```

vdimir replied:
About issue with `join_use_nulls` not sure if it's doable with a pass or not. probably we may have have a pass that change types in such expressions in `JOIN ON`, seems more like workaround, but I don't have an idea hot to fix it properly for the moment.

lgbo-ustc replied:
I have had some idea about the `query pass`, I think could make it in few days, just take a try at first.

lgbo-ustc replied:
About the common code in `joinRightColumns`, I have no more idea at present. If you have time, could you help to think about it?

vdimir commented on src/Interpreters/TableJoin.h:
```diff
@@ -153,6 +154,8 @@ class TableJoin
 
     Clauses clauses;
+    /// Originally used for inequal join. If there is no any inequal join condition, it will be nullptr.
+    std::shared_ptr<ExpressionActions> full_join_expression = nullptr;
```
`full` in join context may confuse, consider renaming it. Can't we also use `mixed_condition` here?

lgbo-ustc replied:
This is missed in the last rename

lgbo-ustc commented:
@vdimir  We make a new query tree pass to handle the case `join_use_nulls=1`, https://github.com/ClickHouse/ClickHouse/pull/62684

lgbo-ustc committed:
disable when join_use_nulls=1

update setting

update

lgbo-ustc commented:
@vdimir I think we could disable this feature when `join_use_nulls=true`, just throw an exception if there is any mixed condition.

alexey-milovidov commented on src/Core/Settings.h:
```diff
@@ -320,6 +320,7 @@ class IColumn;
     M(Bool, fsync_metadata, true, "Do fsync after changing metadata for tables and databases (.sql files). Could be disabled in case of poor latency on server with high load of DDL queries and high load of disk subsystem.", 0)    \
     \
     M(Bool, join_use_nulls, false, "Use NULLs for non-joined rows of outer JOINs for types that can be inside Nullable. If false, use default value of corresponding columns data type.", IMPORTANT) \
+    M(Bool, allow_experimental_join_condition, false, "Support conditions involve columns from both tables in on join expression.", IMPORTANT) \
```
The description looks incomplete.

vdimir commented:
I assume we can leave `allow_experimental_join_condition`, updating description is enough.

I made some small changes on top of the branch, could you please pick a commit https://github.com/ClickHouse/ClickHouse/commit/27a2b1933bc67939869539dbe68cc46888245eb4 

Also, rebasing or squashing current commits would be good, since there're lot's merges with master.

lgbo-ustc committed:
update settting

lgbo-ustc committed:
apply 27a2b19

lgbo-ustc committed:
update

lgbo-ustc committed:
fixed: filter column is a const column

lgbo-ustc commented:
> > AST fuzzer (ubsan) — /build/src/Common/assert_cast.h:50:12: runtime error: downcast of address 0x7fbf4c01e680 which does not point to an object of type 'const... [Details](https://s3.amazonaws.com/clickhouse-test-reports/60920/76fd7fcc72965a26a237f7d099d1455c4ff910fd/ast_fuzzer__ubsan_.html)
> 
> Is it because of my changes or just because `dynamic_cast` replaced with `assert_cast`? Should we handle low cardinality or something else?

It's a `ColumnConst`.

https://s3.amazonaws.com/clickhouse-test-reports/60920/57bdaba312aaf274d7f3438d1e9acd6187be62b2/ast_fuzzer__debug_/fatal.log

lgbo-ustc committed:
fixed: converting filter column type

vdimir reviewed (approved):
Awesome work on this PR! Thanks for your contribution. Let's go with the current state of the feature and try to include it in upcoming release

vdimir commented:
Github states that there's a conflict, without mentioning a files, though..
<img width="784" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/7023786/5f19e592-701a-4c6e-a0b0-cf8598c92dbf">

lgbo-ustc commented:
> Github states that there's a conflict, without mentioning files, though..
> 
> @lgbo-ustc could you please merge/rebase? Or we can merge #63159 (it contains the same commits with preserved author plus merged master, so this PR will be automatically merged with that)
> 
> <img alt="image" width="784" src="https://private-user-images.githubusercontent.com/7023786/326722696-5f19e592-701a-4c6e-a0b0-cf8598c92dbf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5NTQ5ODEsIm5iZiI6MTcxNDk1NDY4MSwicGF0aCI6Ii83MDIzNzg2LzMyNjcyMjY5Ni01ZjE5ZTU5Mi03MDFhLTRjNmUtYTBiMC1jZjg1OThjOTJkYmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDZUMDAxODAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzY4ZmE2OGRkYTBjYzMxMGY5MTc0Njk3YmEyNWY2ZDg4OTc1NzFmZTg5OWQ1MTc1NWJlNWYwMzQzZTM4MGRmMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-g3taHvTA904H1EjjQd6OznxeBMpVA-aU3wdhqFaP2w">

Sorry for the late reply，I'm on vacation these days

vdimir commented:
> Sorry for the late reply，I'm on vacation these days

No worries, now feature is already merged with experimental flag

Nyakutka commented:
Hi, could anybody tell me when `allow_experimental_join_condition` setting will be in production, not experimental?

alexey-milovidov commented:
@Nyakutka, first of all, this setting is unneeded, and will be removed.
Everything should work by default. @vdimir will take care of it.
